src/adversarial/neurify.jl

```diff
@@ -0,0 +1,306 @@
+    # It cannot be used to create a halfspace constraint.
+    # NOTE: how can this come about, and does it mean anything?
+    if !iszero(aL)
+        intervals = intervals .∩ [HalfSpace(aL, -bL), HalfSpace(aL, -bL), HalfSpace(-aL, bL)]
+    end
+    if !iszero(aU)
+        intervals = intervals .∩ [HalfSpace(aU, -bL), HalfSpace(aU, -bL), HalfSpace(-aU, bL)]
+    end
+    return filter(!isempty, intervals)
+end
+
+
+function get_max_nodewise_influence(solver::Neurify,
+                                    nnet::Network,
+                                    reach::SymbolicIntervalGradient,
+                                    max_violation_con::AbstractVector{Float64},
+                                    splits::Vector)
+
+    LΛ, UΛ, radii = reach.LΛ, reach.UΛ, reach.r
+    is_ambiguous_activation(i, j) = (0 < LΛ[i][j] < 1) || (0 < UΛ[i][j] < 1)
+
+    # We want to find the node with the largest influence
+    # Influence is defined as gradient * interval width
+    # The gradient is with respect to a loss defined by the most violated constraint.
+    LG = copy(max_violation_con)
+    UG = copy(max_violation_con)
+
+    i_max, j_max, influence_max = 0, 0, -Inf
+
+    for i in reverse(1:length(nnet.layers))
+        layer = nnet.layers[i]
+        if layer.activation isa ReLU
+            for j in 1:n_nodes(layer)
+                if is_ambiguous_activation(i, j)
+                    # taking `influence = max_gradient * reach.r[i][j]*k` would be
+                    # different from original paper, but can improve the split efficiency.
+                    # where `k = n-i+1`, i.e. counts up from 1 as you go back in layers.
+                    influence = max(abs(LG[j]), abs(UG[j])) * radii[i][j]
+                    if influence >= influence_max && (i, j, influence) ∉ splits
+                        i_max, j_max, influence_max = i, j, influence
+                    end
+                end
+            end
+        end
+
+        LG_hat = max.(LG, 0.0) .* LΛ[i] .+ min.(LG, 0.0) .* UΛ[i]
+        UG_hat = min.(UG, 0.0) .* LΛ[i] .+ max.(UG, 0.0) .* UΛ[i]
+
+        LG, UG = interval_map(layer.weights', LG_hat, UG_hat)
+    end
+
+    # NOTE can omit this line in the paper version
+    i_max == 0 || j_max == 0 && error("Can not find valid node to split")
+
+    push!(splits, (i_max, j_max, influence_max))
+
+    return (i_max, j_max, influence_max)
+end
+
+function forward_layer(solver::Neurify, layer::Layer, input)
+    return forward_act(forward_linear(solver, input, layer), layer)
+end
+
+# Symbolic forward_linear for the first layer
+function forward_linear(solver::Neurify, input::AbstractPolytope, layer::Layer)
+    (W, b) = (layer.weights, layer.bias)
+    sym = SymbolicInterval(hcat(W, b), hcat(W, b), input)
+    LΛ = Vector{Vector{Int64}}(undef, 0)
+    UΛ = Vector{Vector{Int64}}(undef, 0)
+    r = Vector{Vector{Int64}}(undef, 0)
+    return SymbolicIntervalGradient(sym, LΛ, UΛ, r)
+end
```

Comment:
To clarify, this method acts on the input set (and does nothing), and then feeds a `SymbolicIntervalGradient` into `forward_act`. Therefor the first layer seems not to go through a `forward_linear`, step. Is that right?

Reply:
@changliuliu @Wei-TianHao

Reply:
It does go through the first layer, `sym` is computed using the parameters in the first layer.

Reply:
I see what's going on now (the same happens in ReluVal as well, and both algorithms are currently correct, too, so there is no question about that). I guess the "most technically correct" thing to do would be to initialize Up and Low as `[I zeros(N)]` and then send them through a normal forward_linear step to turn that into `[W b]`...

Reply:
Yes, that may be a better way.